src/main/java/com/hazelcast/aws/impl/AwsOperation.java

```diff
@@ -0,0 +1,332 @@
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Scanner;
+import java.util.TimeZone;
+import java.util.concurrent.Callable;
+import java.util.concurrent.TimeUnit;
+
+import static com.hazelcast.aws.utility.MetadataUtil.IAM_SECURITY_CREDENTIALS_URI;
+import static com.hazelcast.aws.utility.MetadataUtil.INSTANCE_METADATA_URI;
+import static com.hazelcast.aws.utility.StringUtil.isEmpty;
+import static com.hazelcast.aws.utility.StringUtil.isNotEmpty;
+import static com.hazelcast.nio.IOUtil.closeResource;
+
+/**
+ *
+ */
+public abstract class AwsOperation<E> {
+
+    /**
+     * URI to fetch container credentials (when IAM role is enabled)
+     * <p>
+     * see http://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-iam-roles.html
+     */
+    static final String IAM_TASK_ROLE_ENDPOINT = "http://169.254.170.2";
+    private static final ILogger LOGGER = Logger.getLogger(AwsOperation.class);
+    private static final int MIN_HTTP_CODE_FOR_AWS_ERROR = 400;
+    private static final int MAX_HTTP_CODE_FOR_AWS_ERROR = 600;
+    private static final String UTF8_ENCODING = "UTF-8";
+
+    protected AwsConfig awsConfig;
+    protected AwsCredentials awsCredentials;
+    protected URL endpointURL;
+    protected Map<String, String> attributes = new HashMap<String, String>();
+    protected Map<String, String> headers = new HashMap<String, String>();
+    protected String body;
+
+    protected final String service;
+    protected final String docVersion;
+    private final String httpMethod;
+
+    protected AwsOperation(AwsConfig awsConfig, URL endpointURL, String service, String docVersion, String httpMethod) {
+        this.awsConfig = awsConfig;
+        this.awsCredentials = new AwsCredentials(awsConfig);
+        this.endpointURL = endpointURL;
+        this.service = service;
+        this.docVersion = docVersion;
+        this.httpMethod = httpMethod;
+    }
+
+    protected boolean isEcsEndpoint() {
+        return endpointURL.getAuthority().startsWith("ecs.");
+    }
+
+    /**
+     * Invokes this service, unmarshal the response and return it.
+     *
+     * @return the response
+     * @throws Exception if there is an exception invoking the service
+     */
+    public E execute(Object... args) throws Exception {
+        if (isNotEmpty(awsCredentials.getIamRole()) || isEmpty(awsCredentials.getAccessKey())) {
+            retrieveCredentials();
+        }
+
+        LOGGER.finest("OK we have credentials, signing request...");
+
+        prepareHttpRequest(args);
+
+        Aws4RequestSigner requestSigner = getRequestSigner();
+        requestSigner.sign(attributes, headers, body, httpMethod);
+        headers.put("Authorization", requestSigner.getAuthorizationHeader());
+        String securityToken = awsCredentials.getSecurityToken();
+        if (StringUtil.isNotEmpty(securityToken)) {
+            headers.put("X-Amz-Security-Token", securityToken);
+        }
+
```

Comment:
There was such a field we were setting previously`"X-Amz-Signature`. Aren't we using it anymore?

Reply:
Exactly, we don't use it because that value is now passed as a part of the `Authorization` header